### PR TITLE
BDOG-705 user revoke consent (part 3)

### PR DESCRIPTION
**Pull request by natcarey**
Part 3 adds the functionality for the user to see their preferences played back on the consent page and save any updates they make to it.

What's missing from this is:

 - Confirmation on save
 - Reading the consent preference and passing it into GTM

### CODE REVIEW

File: app/uk/gov/hmrc/trackingconsentfrontend/assets/package.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
     "build": "webpack --mode production",
     "lint": "stylelint \"**/*.scss\" && standardx **/*.ts",
     "fixlint": "stylelint --fix \"**/*.scss\" && standardx --fix **/*.ts",
-    "test": "npm run lint && jest --collectCoverage",
+    "test": "jest --collectCoverage && npm run lint",
```
Comment (natcarey) [inline app/uk/gov/hmrc/trackingconsentfrontend/assets/package.json]: I've swapped these to allow tests to run regardless of whether linting has passed.  I hit this use case a number of times before changing this - particularly while commenting out code to see which tests break and being forced to remove unused imports before seeing the test results.

File: app/uk/gov/hmrc/trackingconsentfrontend/assets/src/castToArray.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+const castToArray = nodeList => Array.prototype.slice.apply(nodeList)
+
+export default castToArray
```
Comment (natcarey) [inline app/uk/gov/hmrc/trackingconsentfrontend/assets/src/castToArray.ts]: I know @matthewmascord has a different solution to this.  We should standardise after one/both have hit master.

File: app/uk/gov/hmrc/trackingconsentfrontend/assets/test/fixtures/settingsForm.html [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (cjwebb) [inline app/uk/gov/hmrc/trackingconsentfrontend/assets/test/fixtures/settingsForm.html]: What's going on with all the whitespace?

File: app/uk/gov/hmrc/trackingconsentfrontend/assets/src/settingsFormHandler.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  document.addEventListener('DOMContentLoaded', () => {
+    userPreferences.getPreferences()
+    castToArray(document.querySelectorAll('[data-module="cookie-settings"]')).forEach(cookieSettingsForm => {
+      const onValue = cookieSettingsForm.getAttribute('data-on-value')
+      const offValue = cookieSettingsForm.getAttribute('data-off-value')
+      const currentSettings = userPreferences.getPreferences()
+      if (currentSettings) {
+        Object.keys(currentSettings).forEach(setting => {
+          const isOn = currentSettings[setting]
+          cookieSettingsForm.querySelectorAll('input[name=' + setting + '][value=' + (isOn ? onValue : offValue) + ']').forEach(option => {
```
Comment (matthewmascord) [inline app/uk/gov/hmrc/trackingconsentfrontend/assets/src/settingsFormHandler.ts]: Replace with callIfNotNull and querySelector for compatibility with IE10.

File: app/uk/gov/hmrc/trackingconsentfrontend/assets/src/settingsFormHandler.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+            option.checked = true
+          })
+        })
+      }
+
+      cookieSettingsForm.addEventListener('submit', (e) => {
+        e.preventDefault()
+        userPreferences.setPreferences({
+          usage: !!cookieSettingsForm.querySelector('input[name=usage][value=' + onValue + ']').checked,
```
Comment (matthewmascord) [inline app/uk/gov/hmrc/trackingconsentfrontend/assets/src/settingsFormHandler.ts]: Use ES6 template literals here.

File: app/uk/gov/hmrc/trackingconsentfrontend/assets/src/settingsFormHandler.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import castToArray from './castToArray'
+
+const settingsFormHandler = (document, userPreferences) => {
+  document.addEventListener('DOMContentLoaded', () => {
```
Comment (matthewmascord) [inline app/uk/gov/hmrc/trackingconsentfrontend/assets/src/settingsFormHandler.ts]: We should listen for this in the entrypoint module for consistency with the other bundle. Or move both.

File: app/uk/gov/hmrc/trackingconsentfrontend/assets/src/settingsFormHandler.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    userPreferences.getPreferences()
+    castToArray(document.querySelectorAll('[data-module="cookie-settings"]')).forEach(cookieSettingsForm => {
```
Comment (matthewmascord) [inline app/uk/gov/hmrc/trackingconsentfrontend/assets/src/settingsFormHandler.ts]: Replace with callIfNotNull and querySelector for compatibility with IE10.

File: app/uk/gov/hmrc/trackingconsentfrontend/assets/src/settingsFormHandler.ts (hunk 6 above)
Comment (natcarey) [reply app/uk/gov/hmrc/trackingconsentfrontend/assets/src/settingsFormHandler.ts]: What's the test strategy for that?

File: app/uk/gov/hmrc/trackingconsentfrontend/assets/test/fixtures/settingsForm.html (hunk 3 above)
Comment (natcarey) [reply app/uk/gov/hmrc/trackingconsentfrontend/assets/test/fixtures/settingsForm.html]: It was taken from a rendered version of the form, I didn't change what was provided to the browser.

**Pull request merged**